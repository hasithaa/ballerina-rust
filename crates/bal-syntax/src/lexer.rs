//! Lexer implementation for Ballerina

use logos::Logos;

#[derive(Debug, Clone, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
    pub line: usize,    // 1-based line number
    pub column: usize,  // 0-based column
}

#[derive(Debug, Clone, PartialEq)]
pub struct TokenInfo {
    pub kind: Token,
    pub text: String,
    pub span: Span,
}

#[derive(Logos, Debug, PartialEq, Clone)]
#[logos(skip r"[ \t\n\f]+")]  // Ignore whitespace
pub enum Token {
    // Keywords
    #[token("import")]
    Import,
    
    #[token("public")]
    Public,
    
    #[token("function")]
    Function,
    
    #[token("returns")]
    Returns,
    
    #[token("int")]
    Int,
    
    #[token("boolean")]
    Boolean,
    
    #[token("if")]
    If,
    
    #[token("else")]
    Else,
    
    #[token("while")]
    While,
    
    #[token("break")]
    Break,
    
    #[token("continue")]
    Continue,
    
    #[token("return")]
    Return,
    
    #[token("true")]
    True,
    
    #[token("false")]
    False,
    
    // Operators
    #[token("==")]
    Eq,
    
    #[token("!=")]
    NotEq,
    
    #[token("<")]
    Lt,
    
    #[token("<=")]
    LtEq,
    
    #[token(">")]
    Gt,
    
    #[token(">=")]
    GtEq,
    
    #[token("+")]
    Plus,
    
    #[token("-")]
    Minus,
    
    #[token("*")]
    Star,
    
    #[token("/")]
    Slash,
    
    #[token("%")]
    Percent,
    
    #[token("=")]
    Assign,
    
    // Delimiters
    #[token("(")]
    LParen,
    
    #[token(")")]
    RParen,
    
    #[token("{")]
    LBrace,
    
    #[token("}")]
    RBrace,
    
    #[token(",")]
    Comma,
    
    #[token(":")]
    Colon,
    
    #[token(";")]
    Semicolon,
    
    // Identifiers and literals
    #[regex("[A-Za-z][A-Za-z0-9_]*")]
    Identifier,
    
    #[regex("0|[1-9][0-9]*")]
    IntegerLiteral,

    // Comments
    #[regex("//[^\n]*")]
    LineComment,
}

pub struct Lexer<'a> {
    inner: logos::Lexer<'a, Token>,
    source: &'a str,
    line_starts: Vec<usize>, // Cache line start positions
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        // Calculate line starts once during initialization
        let line_starts: Vec<_> = std::iter::once(0)
            .chain(input.match_indices('\n')
                .map(|(i, _)| i + 1))
            .collect();
        
        Self {
            inner: Token::lexer(input),
            source: input,
            line_starts,
        }
    }

    fn get_position(&self, offset: usize) -> (usize, usize) {
        match self.line_starts.binary_search(&offset) {
            Ok(line) => (line + 1, 0),
            Err(line) => (line, offset - self.line_starts[line - 1]),
        }
    }

    pub fn next_token(&mut self) -> Option<TokenInfo> {
        self.inner.next().map(|token| {
            let range = self.inner.span();
            let (line, column) = self.get_position(range.start);
            TokenInfo {
                kind: token.unwrap(),
                text: self.source[range.clone()].to_string(),
                span: Span {
                    start: range.start,
                    end: range.end,
                    line,
                    column,
                },
            }
        })
    }
} 