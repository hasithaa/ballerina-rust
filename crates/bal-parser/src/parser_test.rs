#[cfg(test)]
mod tests {
    use crate::Parser;
    use bal_syntax::SyntaxKind;

    fn parse(input: &str) -> String {
        let tokens = tokenize(input);
        let parser = Parser::new(tokens);
        let node = parser.parse();
        format!("{:#?}", node)
    }

    fn tokenize(input: &str) -> Vec<(SyntaxKind, String)> {
        use bal_syntax::lexer::{Lexer, Token};
        
        let mut lexer = Lexer::new(input);
        let mut tokens = Vec::new();
        
        while let Some(token_info) = lexer.next_token() {
            let kind = match token_info.kind {
                Token::Import => SyntaxKind::IMPORT_KW,
                Token::Public => SyntaxKind::PUBLIC_KW,
                Token::Function => SyntaxKind::FUNCTION_KW,
                Token::Returns => SyntaxKind::RETURNS_KW,
                Token::Int => SyntaxKind::INT_KW,
                Token::Boolean => SyntaxKind::BOOLEAN_KW,
                Token::Identifier => SyntaxKind::IDENTIFIER,
                Token::LParen => SyntaxKind::L_PAREN,
                Token::RParen => SyntaxKind::R_PAREN,
                Token::LBrace => SyntaxKind::L_BRACE,
                Token::RBrace => SyntaxKind::R_BRACE,
                Token::Comma => SyntaxKind::COMMA,
                Token::Semicolon => SyntaxKind::SEMICOLON,
                Token::Slash => SyntaxKind::SLASH,
                _ => continue, // Skip other tokens for now
            };
            tokens.push((kind, token_info.text));
        }
        tokens
    }

    #[test]
    fn test_empty_file() {
        let input = "";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_import() {
        let input = "import foo/bar;";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_function() {
        let input = "function add(int a, int b) returns int { }";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_public_function() {
        let input = "public function greet() { }";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_function_with_multiple_params() {
        let input = "function calc(int x, boolean flag, int y) { }";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_complete_module() {
        let input = r#"
            import math/basic;
            
            public function add(int a, int b) returns int {
            }
            
            function helper(boolean flag) {
            }
        "#;
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_error_recovery_missing_semicolon() {
        let input = "import foo/bar";  // Missing semicolon
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_error_recovery_missing_param_type() {
        let input = "function bad(a, int b) { }";  // Missing type for first param
        insta::assert_snapshot!(parse(input));
    }
} 