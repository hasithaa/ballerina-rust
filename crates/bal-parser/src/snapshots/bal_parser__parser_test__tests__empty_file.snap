---
source: crates/bal-parser/src/parser_test.rs
assertion_line: 45
expression: parse(input)
---
GreenNode {
    kind: SyntaxKind(
        38,
    ),
    text_len: 0,
    n_children: 0,
}
